src/v3/webpack.dev.config.ts

```diff
@@ -0,0 +1,136 @@
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import fs from 'fs';
+import { resolve } from 'path';
+
+import nodemon from 'nodemon';
+import webpack from 'webpack';
+import { merge } from 'webpack-merge';
+import MiniCssExtractPlugin from 'mini-css-extract-plugin';
+
+import baseConfig from './webpack.common.config';
+
+import type { Configuration } from 'webpack';
+import type { Configuration as DevServerConfiguration } from 'webpack-dev-server';
+
+const DEV_SERVER_PORT = 3000;
+const MOCK_SERVER_PORT = 3030;
+
+const WIDGET_RC_JS = resolve(__dirname, '../..', '.widgetrc.js');
+const PLAYGROUND = resolve(__dirname, '../..', 'playground');
+const TARGET = resolve(__dirname, '../..', 'target');
+const ASSETS = resolve(__dirname, '../..', 'assets');
+
+// run `OKTA_SIW_HOST=0.0.0.0 yarn start --watch` to override the host
+const HOST = process.env.OKTA_SIW_HOST || 'localhost';
+const STATIC_DIRS = [PLAYGROUND, TARGET, ASSETS];
+
+const headers = (() => {
+  if (!process.env.DISABLE_CSP) {
+    // Allow google domains for testing recaptcha
+    const scriptSrc = `script-src http://${HOST}:${DEV_SERVER_PORT} https://www.google.com https://www.gstatic.com`;
+    const styleSrc = `style-src http://${HOST}:${DEV_SERVER_PORT} 'nonce-playground'`;
+
+    return {
+      'Content-Security-Policy': `${scriptSrc}; ${styleSrc};`,
+    };
+  }
+  return;
+})();
+
+if (!fs.existsSync(WIDGET_RC_JS)) {
+  // create default WIDGET_RC if it doesn't exist to simplifed the build process
+  fs.copyFileSync(resolve(__dirname, '../..', '.widgetrc.sample.js'), WIDGET_RC_JS);
+}
+
+const devConfig: Configuration = merge(
+  baseConfig,
+  {
+    mode: 'development',
+    devtool: 'source-map',
+    entry: {
+      playground: {
+        import: [
+          `${PLAYGROUND}/main.ts`,
+        ],
+        filename: 'playground.bundle.js',
+      },
+      widget: {
+        import: [
+          // polyfill must appear first in entry point array
+          resolve(__dirname, '../..', './polyfill/modern.js'),
+          resolve(__dirname, 'src/index.ts'),
+        ],
+        filename: 'js/okta-sign-in.js',
+        library: {
+          name: 'OktaSignIn',
+          type: 'umd',
+          export: 'default',
+        },
+      },
+    },
+    output: {
+      path: `${PLAYGROUND}/target`,
+    },
+    resolve: {
+      alias: {
+        '@okta/duo': `${PLAYGROUND}/mocks/spec-duo/duo-mock.js`,
+        duo_web_sdk: resolve(__dirname, 'src/__mocks__/duo_web_sdk'),
+      }
+    },
+    plugins: [
+      new MiniCssExtractPlugin({
```

Comment:
move to common?

Reply:
Since the options passed to the plugin constructor are different (different filename) it's much easier to keep them individually in separate files and it makes it clearer which instance of the plugin will be in the result. This is why we never actually use the `webpack.common.config.ts` file as an input to `webpack-cli` and also why I define the export type as `Partial<Configuration>` rather than `Configuration`.